### PR TITLE
Make `coerce` working within `items` rule

### PR DESCRIPTION
**Pull request by oev81**
Case example:
`schema = {'things': {'type': 'list', 'items': [{'coerce': int}, {'coerce': str}]}}`

### CODE REVIEW

File: cerberus/validator.py [hunk 1]
```diff
@@ -747,6 +749,19 @@ def __normalize_sequence(self, field, mapping, schema):
             self._drop_nodes_from_errorpaths(validator._errors, [], [2])
             self._error(validator._errors)
 
+    def __normalize_items(self, field, mapping, schema):
+        schema = dict(((k, v) for k, v in enumerate(schema[field]['items'])))
+        document = dict((k, v) for k, v in enumerate(mapping[field]))
+        validator = self._get_child_validator(
+            document_crumb=field, schema_crumb=(field, 'items'),
+            schema=schema)
+        value_type = type(mapping[field])
```
[the remark this answers is not in the document]
Comment (oev81) [reply cerberus/validator.py]: Hi!
> i'd like to see a test that covers the case when a value in a document throws a ValueError on coerce.

I've added tests for `ValueError` and `TypeError`.

> why is `type` being called here?

I used `__normalize_sequence` as a base. This line is added by commit eb24176 (Ensures container types are preserved in normalization). If `tuple` passed instead `list`, type will be preserved.

```
>>> schema = {'things': {'type': 'list', 'items': [{'coerce': int}, {'coerce': str}]}}
>>> validator = Validator(schema)
>>> validator.validate({'things': ['1', 2]})
True
>>> validator.document
{'things': [1, '2']}
>>> validator.validate({'things': ('1', 2)})
True
>>> validator.document
{'things': (1, '2')}
```

**Comment by funkyfuture** [comment]
a'ight, cool.

> I used __normalize_sequence as a base.

ah, thanks for the reminder. ;-)

just one thing, that should be a simple: when the number of fields in a sequence of a documnet is different from the amount of defined items in the relating `items` constraint, no normalization should be applied as this part of the document will [fail](https://github.com/pyeve/cerberus/blob/master/cerberus/validator.py#L1051) later. does this make sense?

**Comment by oev81** [comment]
> does this make sense?

Well... It seems I understand now why you asking :)
I'm playing with `cerberus` less than a week. I'm learning its features while making a prototype for some real API validation. And that is how I have run into case, when `coerce` not working. I searched a way to convert some parameter value within a validation, and the only way I found is `coerce` rule. This rule is applied before validation, and now I've understood that it works only for simple cases, and doesn't work for cases with rules like  `oneof_schema` or `oneof_items`, or so.

So, is there alternative to `coerce`? Some workaround?

Answer for your question is, yes, lengths must match :)

**Comment by funkyfuture** [comment]
now, i don't get your question. due to lack of context. but i'd recommend StackOverflow for such questions.

regarding the length match; there's just a condition missing in line 688 to address this, right?

oh, and one possibility isn't covered: when an `items` and a `schema` rule are defined on a field. this may not be common but is valid. it's not hard to change the logic, but the question is in which order these would be applied. both possible orders would make sense in different contexts. :-/

**Comment by funkyfuture** [comment]
@oev81 are you going to finalize this patch?

**Comment by oev81** [comment]
> are you going to finalize this patch?

@funkyfuture, sorry for not answering. I found the approach of coerce (conversion before validation) does not feet my task (conversion after validation, or conversion within validation). 
I'm not interested in the patch now. 
Please, drop the patch, if you don't need it.

**Comment by funkyfuture** [comment]
@nicolaiarocci would you re-open this pr and assign me to it? i'd complete this blind spot some day.

**Pull request closed without merging**